File: src/session-recency.ts
import type { SessionUserLabel } from "./session-semantics.js";

/**
 * The single source of truth for an agent's "time anchor": which timestamp to show
 * as relative recency and what verb describes it ("output", "prompted", "idle", …).
 * Both the dashboard and Exposé derive their recency from this so they never disagree.
 *
 * Pending/lifecycle states (starting/stopping/…) are handled by the caller, since the
 * value selection there is runtime-specific.
 */
export interface SessionRecencyInput {
  label?: SessionUserLabel;
  latestUnreadAt?: string;
  lastOutputAt?: string;
  becameIdleAt?: string;
  lastUsedAt?: string;
}

export interface SessionRecencyAnchor {
  label: string;
  value?: string;
}

export function sessionRecencyAnchor(input: SessionRecencyInput): SessionRecencyAnchor | null {
  const { label, latestUnreadAt, lastOutputAt, becameIdleAt, lastUsedAt } = input;
  const output = lastOutputAt ? { label: "output", value: lastOutputAt } : null;
  switch (label) {
    case "needs_input":
    case "needs_response":
      return { label: "prompted", value: latestUnreadAt ?? lastOutputAt ?? becameIdleAt ?? lastUsedAt };
    case "next_step":
    case "idle":
    case "interrupted":
      return output ?? { label: "idle", value: becameIdleAt ?? lastUsedAt };
    case "working":
    case "ready":
      return output;
    case "done":
      return output ?? { label: "done", value: becameIdleAt ?? lastUsedAt };
    case "offline":
      return output ?? { label: "offline", value: lastUsedAt };
    case "blocked":
      return { label: "blocked", value: latestUnreadAt ?? becameIdleAt ?? lastOutputAt ?? lastUsedAt };
    case "error":
      return { label: "failed", value: latestUnreadAt ?? becameIdleAt ?? lastOutputAt ?? lastUsedAt };
    default:
      return output;
  }
}

File: native/crates/aimux/src/session_recency.rs
use serde_json::{Value, json};

pub fn session_recency_anchor(input: &Value) -> Value {
    let label = input.get("label").and_then(Value::as_str);
    let latest_unread_at = input.get("latestUnreadAt").and_then(Value::as_str);
    let last_output_at = input.get("lastOutputAt").and_then(Value::as_str);
    let became_idle_at = input.get("becameIdleAt").and_then(Value::as_str);
    let last_used_at = input.get("lastUsedAt").and_then(Value::as_str);
    let output = last_output_at.map(|value| json!({ "label": "output", "value": value }));
    match label {
        Some("needs_input" | "needs_response") => json!({
            "label": "prompted",
            "value": latest_unread_at.or(last_output_at).or(became_idle_at).or(last_used_at),
        }),
        Some("next_step" | "idle" | "interrupted") => output.unwrap_or_else(
            || json!({ "label": "idle", "value": became_idle_at.or(last_used_at) }),
        ),
        Some("working" | "ready") => output.unwrap_or(Value::Null),
        Some("done") => output.unwrap_or_else(
            || json!({ "label": "done", "value": became_idle_at.or(last_used_at) }),
        ),
        Some("offline") => {
            output.unwrap_or_else(|| json!({ "label": "offline", "value": last_used_at }))
        }
        Some("blocked") => json!({
            "label": "blocked",
            "value": latest_unread_at.or(became_idle_at).or(last_output_at).or(last_used_at),
        }),
        Some("error") => json!({
            "label": "failed",
            "value": latest_unread_at.or(became_idle_at).or(last_output_at).or(last_used_at),
        }),
        _ => output.unwrap_or(Value::Null),
    }
}
